_posts/2024-05-07-opentelemetry-metrics-visualization.md

```diff
@@ -0,0 +1,421 @@
+        sslKeyFile: "./key.pem"
+    processor:
+    - otel_metrics:
+    sink:
+    - opensearch:
+        hosts: [ "<your-OpenSearch-url>" ]
+        username: "otelWriter"
+        password: "<your-password>"
+        index: metrics-otel-v1-%{yyyy.MM.dd}
+```
+
+This example uses Basic Auth for OpenSearch.
+It requires the `otelWriter` backend role with sufficient permissions for DataPrepper.
+This is similar to trace ingestion.
+
+The complete data flow of this setup is shown in the following diagram:
+
+![Data Flow of OTel Ingestion](/assets/media/blog-images/2024-05-07-opentelemetry-metrics-visualization/otel-ingestion-architecture.drawio.png){:class="img-centered"}
+
+When this setup is complete, you will have a lot of metrics in OpenSearch to analyze.
+There is a complete [documentation](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/kubeletstatsreceiver/documentation.md) of the metrics and metadata emitted by the kubeletstatsreceiver.
+
+> Hint: If you are a service provider for development teams, this setup can be automated and transparent to your users.
+Within SAP BTP, this is achieved by the [Kyma Telemetry module](https://kyma-project.io/#/telemetry-manager/user/README) in connection with [SAP Cloud Logging](https://discovery-center.cloud.sap/serviceCatalog/cloud-logging).
+It is this solution, that forms the foundation of this article.
+
+### OpenTelemetry Metrics Data Model
+
+OpenTelemetry uses a well documented [Metrics Data Model](https://opentelemetry.io/docs/specs/otel/metrics/data-model/).
+The linked article explains all the details of the various supported metric types.
+We will do a less thorough approach and inspect example points to collect the necessary information.
+With our example setup, we can find data points similar to this (shortened) example:
+
+```json
+{
+  "_source": {
+    "name": "container.cpu.time",
+    "description": "Total cumulative CPU time (sum of all cores) spent by the container/pod/node since its creation",
+    "value": 120.353905,
+    "flags": 0,
+    "unit": "s",
+    "startTime": "2024-04-10T06:49:56Z",
+    "time": "2024-05-03T07:50:22.197391185Z",
+    "kind": "SUM",
+    "aggregationTemporality": "AGGREGATION_TEMPORALITY_CUMULATIVE",
+    "isMonotonic": true,
+    "serviceName": "otel-emailservice",
+    "instrumentationScope.name": "otelcol/kubeletstatsreceiver",
+    "instrumentationScope.version": "0.97.0",
+    "resource.attributes.k8s@namespace@name": "example",
+    "resource.attributes.k8s@deployment@name": "otel-emailservice",
+    "resource.attributes.k8s@pod@name": "otel-emailservice-7795877686-r2lrt",
+    "resource.attributes.k8s@container@name": "emailservice",
+    "resource.attributes.service@name": "otel-emailservice"
+  }
+}
+```
+
+We can see, that OpenTelemetry separates metrics by the `name`.
+Each point contains the generic `value` and `time`, which form the basic time series for our visualizations.
```

Comment:
```suggestion
Each point contains the generic `value` and `time`, which form the basic time series for the visualizations.
```
I'd rather use "the" instead of "our", but that's a matter of taste